### PR TITLE
fix(desktop): auto-recover unresponsive host-service with a health watchdog

### PR DESCRIPTION
**Pull request by sasha1107**
## What & why

Related to #5699 — "Failed to run preset / Failed to fetch (127.0.0.1:48135)". (The primary root cause — a node-pty fd leak that exhausts the system-wide pty cap — is fixed in #5714; this PR fixes the secondary failure mode where a wedged host-service is never detected.)

The `HostServiceCoordinator` only listens for the child process **exit** event. A host-service that wedges without dying (blocked event loop, fd exhaustion after extended usage) therefore stays `"running"` forever: `getConnection` keeps handing the renderer the same port, every `terminal.createSession` fetch fails with `Failed to fetch (127.0.0.1:<port>)`, and the only recovery is the manual tray restart — or rebooting the machine, as the reporter did. (The port in the report, 48135, is in the coordinator's stable-port range 48000–48999, confirming the renderer was talking to a local host-service the coordinator still considered healthy.)

This adds a health watchdog to the coordinator:

- Probes every `"running"` instance's `/trpc/health.check` every 15s (3s probe timeout), skipping instances with a pending start and discarding probe results if the instance was stopped/replaced mid-probe.
- After **3 consecutive** failed checks, force-restarts that org's host-service via the existing `reset()` recovery path, escalating to SIGKILL on the wedged pid (a wedged process can ignore `stop()`'s SIGTERM).
- Terminals survive the restart through the pty-daemon adoption path `createSession` already implements; renderer picks up the fresh connection through the existing 5s `getConnection` poll + status events.

Related: #5700 retries transient network blips on the create path; this PR addresses the persistent-until-reboot failure mode by making the coordinator detect and heal a wedged child.

## How I tested it

- Added 6 unit tests in `host-service-coordinator.test.ts` covering: healthy probe clears the failure counter; no restart below threshold; force-restart at threshold with SIGTERM→SIGKILL escalation; skip when no spawn config (logged-out); non-running instances not probed; probe failures don't count against a replaced child.
- `bun test apps/desktop/src/main/lib/host-service-coordinator.test.ts` — 12 pass (6 existing + 6 new).
- `bun run lint` and `tsc --noEmit` (apps/desktop) pass.

## Checklist

- [x] PR title follows conventional commits (`type(scope): subject`)
- [x] `bun run lint` and `bun run typecheck` pass (CI fails on lint warnings too)
- [x] "Allow edits from maintainers" is checked on fork PRs

---

<a href="https://stagereview.app/superset-sh/superset/pull/5711">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

---
## Summary by cubic
Adds a health watchdog to the desktop `HostServiceCoordinator` to auto-recover unresponsive host-service processes. It detects wedged, still-running children and restarts them, removing persistent “Failed to fetch (127.0.0.1:PORT)” errors without manual restarts.

- **Bug Fixes**
  - Probe running, owned instances’ `/trpc/health.check` every 15s (3s timeout); clamp fetch/sleep to the remaining budget; avoid overlapping ticks.
  - Bind failure counts to the child pid; ignore stale probe results; clear on success or deliberate `stop()`.
  - After 3 consecutive failures, force-restart via `reset()` and SIGKILL the wedged pid before respawn; skip adopted (non-owned) instances.
  - Revalidate after config awaits; keep counters armed when config is missing; retry respawn when recovery fails or no instance is tracked without overriding a concurrent `stop()`/start.
  - Remove ghost “starting” entries on spawn errors; watchdog tick logs and swallows its own failures; wire `enableHealthWatchdog` in `main` with a shared auth/config provider; tests cover thresholds, pid-binding, retries, escalation ordering, revalidation, adopted-skip, and startup failure cleanup.

<sup>Written for commit a1d032a2ab901814dae0e7ab152492487fc8c39d. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/superset-sh/superset/pull/5711?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Reliability Improvements**
  * Added a health watchdog that periodically probes running host services and recovers stalled instances by force-resetting them after consecutive failures.
  * Recovery includes retries when no healthy instance is available, and escalates from graceful termination to force-kill; intentional stops now prevent auto-respawn.
  * Improved health-check timing to ensure probes and retries always respect the overall timeout budget.
* **Development Experience**
  * Streamlined development reload by reusing shared auth/API configuration for host-service startup.
* **Tests**
  * Expanded watchdog coverage for thresholds, replacement edge cases, retry behavior, and cleanup/disarm scenarios.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 6 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Refine health check polling budget](https://stagereview.app/superset-sh/superset/pull/5711/chapters/1) |
| 2 | [Define watchdog constants and state](https://stagereview.app/superset-sh/superset/pull/5711/chapters/2) |
| 3 | [Enhance process recovery and cleanup](https://stagereview.app/superset-sh/superset/pull/5711/chapters/3) |
| 4 | [Implement health watchdog logic](https://stagereview.app/superset-sh/superset/pull/5711/chapters/4) |
| 5 | [Wire watchdog into application lifecycle](https://stagereview.app/superset-sh/superset/pull/5711/chapters/5) |
| 6 | [Verify watchdog with unit tests](https://stagereview.app/superset-sh/superset/pull/5711/chapters/6) |

<a href="https://stagereview.app/superset-sh/superset/pull/5711">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 9283ccd9ff41b3c6365f4209c5734f0c9e07f30a on Jul 20, 2026 2:02am UTC.</sub>

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 1]
```diff
@@ -348,6 +360,110 @@ export class HostServiceCoordinator extends EventEmitter {
...
+					(this.watchdogFailures.get(organizationId) ?? 0) + 1;
+				this.watchdogFailures.set(organizationId, failureCount);
+				log.warn(
+					`[host-service:${organizationId}] health check failed (${failureCount}/${HEALTH_WATCHDOG_FAILURE_THRESHOLD}) on port ${instance.port}`,
+				);
+				if (failureCount < HEALTH_WATCHDOG_FAILURE_THRESHOLD) continue;
+
+				this.watchdogFailures.delete(organizationId);
+				const config = await configProvider();
+				if (!config) continue;
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:640]: Fixed in 466695177 — the tracked instance is now revalidated (same pid, still "running", no pending start) after the `configProvider()` await, so a child replaced by a concurrent manual restart is never reset by the watchdog. Covered by the new test "skips the restart when the child is replaced while fetching config".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 2]
```diff
@@ -348,6 +360,110 @@ export class HostServiceCoordinator extends EventEmitter {
...
+
+		try {
+			for (const [organizationId, instance] of [...this.instances]) {
+				if (
+					instance.status !== "running" ||
+					this.pendingStarts.has(organizationId)
+				) {
+					this.watchdogFailures.delete(organizationId);
+					continue;
+				}
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts]: Fixed in 466695177 — the counter is now only cleared after `reset()` succeeds (or when the child turns out to be stale). A null config (logged out) keeps the watchdog armed, so recovery runs on the very next tick once a token is available. Test updated to assert the counter survives and the next tick restarts immediately.

File: apps/desktop/src/main/lib/host-service-coordinator.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts]: Fixed in 466695177 — the counter now survives a null config and is only cleared once recovery actually runs, so the watchdog retries on the next tick instead of waiting out another full threshold.

File: apps/desktop/src/main/index.ts [hunk 3]
```diff
@@ -417,12 +417,14 @@ if (!gotTheLock) {
 			console.error("[main] Failed to install bundled CLI shim:", error);
 		}
 
+		const hostServiceConfigProvider = async () => {
+			const { token } = await loadToken();
+			if (!token) return null;
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/index.ts:446]: Leaving as-is intentionally: this mirrors the existing `enableDevReload` call site, which also discards its disposer. The coordinator is a process-lifetime singleton and the timer is `unref()`d, so there's no scope it can outlive today. Happy to wire both disposers into a teardown if a maintainer prefers.

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 4]
```diff
@@ -348,6 +360,128 @@ export class HostServiceCoordinator extends EventEmitter {
...
+					);
+					continue;
+				}
+				this.watchdogFailures.delete(organizationId);
+				// reset() SIGKILLs the pid named by the child's manifest; if the
+				// wedged child never wrote (or already lost) its manifest it only
+				// got stop()'s SIGTERM, which a wedged process can ignore.
+				if (isProcessAlive(wedgedPid)) {
+					try {
+						killProcess(wedgedPid, "SIGKILL");
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts]: Fixed in ca5446c39 — escalation moved inside `reset()` via a new `alsoKillPid` option, so the SIGKILL fires right after `stop()` and before the (slow) respawn; the pid can no longer be recycled in between. Covered by the new test "escalates the wedged pid to SIGKILL before the replacement spawns".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 5]
```diff
@@ -348,6 +360,128 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				// Revalidate after the await — a manual restart may have replaced
+				// the child while the config was being fetched.
+				const latest = this.instances.get(organizationId);
+				if (
+					!latest ||
+					latest.pid !== instance.pid ||
+					latest.status !== "running" ||
+					this.pendingStarts.has(organizationId)
+				) {
+					this.watchdogFailures.delete(organizationId);
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:651]: Fixed in ca5446c39 — `reset()` now takes `alsoKillPid` and performs the SIGKILL escalation immediately after `stop()`, before spawning the replacement. A new test asserts the SIGKILL is observed before spawn runs.

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 6]
```diff
@@ -348,6 +360,128 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				try {
+					await this.reset(organizationId, config);
+				} catch (error) {
+					// Counter intentionally kept — the next failed probe pushes it
+					// past the threshold again and recovery is retried.
+					log.error(
+						`[host-service:${organizationId}] watchdog restart failed:`,
+						error,
+					);
+					continue;
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:690]: Fixed in ca5446c39 — the tick now has a second pass that retries armed orgs with no tracked instance, so a failed respawn is re-attempted every tick (bounded by the 15s cadence) until one sticks. `stop()` disarms the entry so deliberate stops are never auto-respawned. Tests: "a failed recovery is retried until a spawn sticks" and "stays armed while respawns keep failing".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 7]
```diff
@@ -348,6 +360,128 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				}
+
+				log.error(
+					`[host-service:${organizationId}] unresponsive for ${HEALTH_WATCHDOG_FAILURE_THRESHOLD} consecutive health checks — force-restarting`,
+				);
+				const wedgedPid = instance.pid;
+				try {
+					await this.reset(organizationId, config);
+				} catch (error) {
+					// Counter intentionally kept — the next failed probe pushes it
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts]: Fixed in ca5446c39 — valid catch, the promised retry indeed never ran once the instance was gone. The tick now retries armed instance-less orgs each pass; see the two new retry tests.

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 8]
```diff
@@ -91,6 +100,9 @@ export class HostServiceCoordinator extends EventEmitter {
 	private scriptPath = path.join(__dirname, "host-service.js");
 	private machineId = getHostId();
 	private devReloadWatcher: fs.FSWatcher | null = null;
+	private watchdogTimer: ReturnType<typeof setInterval> | null = null;
+	private watchdogTickInFlight = false;
+	private watchdogFailures = new Map<string, number>();
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts]: Fixed in ca5446c39 — counters are now `{ pid, count }` and only increment when the probed child's pid matches, so a replacement child starts from 1. `stop()` clears the entry as well. Test: "failure counter rebinds when a new child replaces the old one between ticks".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 9]
```diff
@@ -348,6 +360,128 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				// Keep the counter until recovery actually runs — if no config is
+				// available (logged out) the watchdog stays armed and retries on
+				// the next tick instead of waiting out a full threshold again.
+				const config = await configProvider();
+				if (!config) continue;
+
+				// Revalidate after the await — a manual restart may have replaced
+				// the child while the config was being fetched.
+				const latest = this.instances.get(organizationId);
+				if (
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:645]: Fixed in ca5446c39 — counter is pid-bound (`{ pid, count }`) and `stop()` clears it, so a fresh child never inherits its predecessor's failures.

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 10]
```diff
@@ -348,6 +360,128 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				) {
+					this.watchdogFailures.delete(organizationId);
+					continue;
+				}
+
+				const healthy = await pollHealthCheck(
+					`http://127.0.0.1:${instance.port}`,
+					instance.secret,
+					HEALTH_WATCHDOG_PROBE_TIMEOUT_MS,
+				);
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:604]: Fixed in ca5446c39 — `pollHealthCheck` now clamps both the per-fetch abort and the retry sleep to the remaining deadline, so a 3s probe is bounded at ~3s and serial probes across orgs stay well inside the 15s tick.

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 11]
```diff
@@ -348,6 +379,165 @@ export class HostServiceCoordinator extends EventEmitter {
...
+					this.watchdogFailures.delete(organizationId);
+					continue;
+				}
+				const config = await configProvider();
+				if (!config) continue;
+				log.warn(
+					`[host-service:${organizationId}] retrying respawn after failed watchdog recovery`,
+				);
+				try {
+					await this.startWithPreferredPorts(organizationId, config);
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:714]: Fixed in cc93ea800 — the retry pass now revalidates after the config await (entry identity unchanged, no instance, no pending start) before respawning, so a deliberate stop() landing mid-fetch is never overridden. Test: "retry pass aborts when the org is stopped while fetching config".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 12]
```diff
@@ -348,6 +379,165 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				}
+			}
+
+			// Orgs whose recovery failed have no tracked instance and are
+			// invisible to the loop above — keep retrying those each tick until
+			// a spawn sticks. Deliberate stops never land here: stop() disarms
+			// the entry.
+			for (const [organizationId, entry] of [...this.watchdogFailures]) {
+				if (
+					this.instances.has(organizationId) ||
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:687]: Fixed at the root in cc93ea800 — the stale "starting" ghost was a spawn() bug: any startup failure after the instance was registered (e.g. buildEnv rejecting) left the entry behind forever. spawn() now removes its entry and emits a stopped status on every failure path, so the retry-pass guard can keep treating any tracked instance as legitimately occupied. Test: "a spawn failure leaves no ghost 'starting' instance behind".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 13]
```diff
@@ -348,6 +379,165 @@ export class HostServiceCoordinator extends EventEmitter {
...
+				) {
+					continue;
+				}
+				if (entry.count < HEALTH_WATCHDOG_FAILURE_THRESHOLD) {
+					// Stale sub-threshold entry (e.g. the child crashed on its
+					// own after a blip) — nothing to recover, drop it.
+					this.watchdogFailures.delete(organizationId);
+					continue;
+				}
+				const config = await configProvider();
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:698]: Fixed in cc93ea800 — runWatchdogTick now catches and logs its own failures, so a rejecting config lookup (or anything else in the tick) can no longer escape the timer callback as an unhandled rejection. Test: "a rejecting config lookup fails the tick quietly".

File: apps/desktop/src/main/lib/host-service-coordinator.ts [hunk 14]
```diff
@@ -197,21 +219,30 @@ export class HostServiceCoordinator extends EventEmitter {
...
+		// so the pid can't be recycled by the OS in the meantime.
 		const preferredPorts = this.getPreferredPorts(organizationId);
 		const manifestPid = readManifest(organizationId)?.pid;
 
 		this.stop(organizationId);
 
-		if (manifestPid != null && isProcessAlive(manifestPid)) {
+		const pidsToKill = new Set<number>();
+		if (manifestPid != null) pidsToKill.add(manifestPid);
+		if (opts.alsoKillPid != null) pidsToKill.add(opts.alsoKillPid);
```
[the remark this answers is not in the document]
Comment (sasha1107) [reply apps/desktop/src/main/lib/host-service-coordinator.ts:278]: Leaving as-is: between stop()'s SIGTERM and the SIGKILL there is no await — it's a handful of synchronous statements, so the reuse window is microseconds and requires the kernel to reap the child AND recycle the exact pid within that same tick, which macOS/Linux pid allocation doesn't do in practice. The realistic multi-second window (escalating after the respawn's health poll) was the previous round's fix. Binding to process identity beyond pid would need start-time comparison via sysctl — happy to add if maintainers want it, but it seems disproportionate to the residual risk.
